arch/lkl/kernel/threads.c

```diff
@@ -69,30 +72,38 @@ void free_thread_stack(unsigned long *stack)
 
 struct thread_info *_current_thread_info = &init_thread_union.thread_info;
 
+/*
+ * schedule() expects the return of this function to be the task that we
+ * switched away from. Returning prev is not going to work because we are
+ * actually going to return the previous taks that was scheduled before the
+ * task we are going to wake up, and not the current task, e.g.:
+ *
+ * swapper -> init: saved prev on swapper stack is swapper
+ * init -> ksoftirqd0: saved prev on init stack is init
+ * ksoftirqd0 -> swapper: returned prev is swapper
+ */
+static struct task_struct *abs_prev = &init_task;
+
 struct task_struct *__switch_to(struct task_struct *prev,
 				struct task_struct *next)
 {
 	struct thread_info *_prev = task_thread_info(prev);
 	struct thread_info *_next = task_thread_info(next);
-	/*
-	 * schedule() expects the return of this function to be the task that we
-	 * switched away from. Returning prev is not going to work because we
-	 * are actually going to return the previous taks that was scheduled
-	 * before the task we are going to wake up, and not the current task,
-	 * e.g.:
-	 *
-	 * swapper -> init: saved prev on swapper stack is swapper
-	 * init -> ksoftirqd0: saved prev on init stack is init
-	 * ksoftirqd0 -> swapper: returned prev is swapper
-	 */
-	static struct task_struct *abs_prev = &init_task;
 
 	_current_thread_info = task_thread_info(next);
 	_next->prev_sched = prev;
 	abs_prev = prev;
 
 	lkl_ops->sem_up(_next->sched_sem);
-	lkl_ops->sem_down(_prev->sched_sem);
+	if (test_ti_thread_flag(_prev, TIF_SCHED_JB)) {
+		clear_ti_thread_flag(_prev, TIF_SCHED_JB);
+		lkl_ops->jmp_buf_longjmp(&_prev->sched_jb, 1);
+	} else if (test_ti_thread_flag(_prev, TIF_SCHED_EXIT)) {
+		clear_ti_thread_flag(_prev, TIF_SCHED_EXIT);
```

Comment:
This is setup with thread_set_sched_exit() from del_host_task. We need this because we need to call do_exit() which calls schedule() and we want to avoid calling finish_switch twice.